internal_event_handlers.go

```diff
@@ -335,6 +335,40 @@ func (wc *workflowEnvironmentImpl) RequestCancelTimer(timerID string) {
 	wc.logger.Debug("RequestCancelTimer", zap.String(tagTimerID, timerID))
 }
 
+func validateVersion(changeID string, version, minSupported, maxSupported Version) {
+	if version < minSupported {
+		panic(fmt.Sprintf("Workflow code removed support of version %v. "+
+			"for \"%v\" changeID. The oldest supported version is %v",
+			version, changeID, minSupported))
+	}
+	if version > maxSupported {
+		panic(fmt.Sprintf("Workflow code is too old to support version %v "+
+			"for \"%v\" changeID. The maximum supported version is %v",
+			version, changeID, minSupported))
+	}
+}
+
+func (wc *workflowEnvironmentImpl) GetVersion(changeID string, minSupported, maxSupported Version) Version {
+	var version Version
+	if wc.isReplay {
+		var ok bool
+		version, ok = wc.changeVersions[changeID]
+		if !ok {
+			version = DefaultVersion
+		}
+		validateVersion(changeID, version, minSupported, maxSupported)
+		wc.logger.Debug("GetVersion return version from a marker",
+			zap.String(tagChangeID, changeID), zap.Int(tagVersion, int(version)))
+		return version
+	}
+	version = maxSupported
+	wc.decisionsHelper.recordVersionMarker(changeID, version)
+
+	wc.logger.Debug("GetVersion return",
```

Comment:
same here? the more it looks this flag of checking to log looks hacky all over place.